src/main/scala/mesosphere/marathon/core/matcher/base/util/OfferOperationFactory.scala

```diff
@@ -47,78 +47,99 @@ class OfferOperationFactory(
+      val reservedResources = resources.map { resource =>
 
-  def createVolumes(
-    reservationLabels: ReservationLabels,
-    localVolumes: Seq[(DiskSource, LocalVolume)]): Mesos.Offer.Operation = {
-
-    val volumes: Seq[Mesos.Resource] = localVolumes.map {
-      case (source, vol) =>
-        val disk = {
-          val persistence = Mesos.Resource.DiskInfo.Persistence.newBuilder().setId(vol.id.idString)
-          principalOpt.foreach(persistence.setPrincipal)
-
-          // Pod volumes have names, whereas app volumes do not. Since pod persistent volumes are created
-          // in the executor container, a persistent volume name is used as its name on the Mesos end. In this case
-          // all the corresponding volume mounts refer to the volume by its name. On the other hand, in case of apps,
-          // volumes do not have names, and since persistent volumes are not shared in this case, the mount path
-          // is used instead.
-          val name = vol.persistentVolume.name.getOrElse(vol.mount.mountPath)
-
-          val mode = VolumeMount.readOnlyToProto(vol.mount.readOnly)
-          val volume = Mesos.Volume.newBuilder()
-            .setContainerPath(name)
-            .setMode(mode)
-
-          val builder = Mesos.Resource.DiskInfo.newBuilder()
-            .setPersistence(persistence)
-            .setVolume(volume)
-          source.asMesos.foreach(builder.setSource)
-          builder
-        }
-
-        val reservation = Mesos.Resource.ReservationInfo.newBuilder()
+        val reservation = ReservationInfo.newBuilder()
           .setLabels(reservationLabels.mesosLabels)
-        principalOpt.foreach(reservation.setPrincipal)
+          .setPrincipal(principal)
 
-        Mesos.Resource.newBuilder()
-          .setName("disk")
-          .setType(Mesos.Value.Type.SCALAR)
-          .setScalar(Mesos.Value.Scalar.newBuilder().setValue(vol.persistentVolume.persistent.size.toDouble).build())
+        Mesos.Resource.newBuilder(resource)
           .setRole(role)
           .setReservation(reservation)
-          .setDisk(disk)
           .build()
-    }
+      }
 
-    val create = Mesos.Offer.Operation.Create.newBuilder()
-      .addAllVolumes(volumes.asJava)
+      val reserve = Mesos.Offer.Operation.Reserve.newBuilder()
+        .addAllResources(reservedResources.asJava)
+        .build()
 
-    Mesos.Offer.Operation.newBuilder()
-      .setType(Mesos.Offer.Operation.Type.CREATE)
-      .setCreate(create)
-      .build()
+      Mesos.Offer.Operation.newBuilder()
+        .setType(Mesos.Offer.Operation.Type.RESERVE)
+        .setReserve(reserve)
+        .build()
+    }.to[Seq]
+  }
+
+  def createVolumes(
+    reservationLabels: ReservationLabels,
+    localVolumes: Seq[(Option[String], DiskSource, LocalVolume)]): Seq[Mesos.Offer.Operation] = {
+
+    // Mesos requires that there is an operation per each resource provider ID
+    localVolumes.groupBy(_._1).values.map { localVolumes =>
```

Comment:
I personally don't like _._1 - try to use case instead to make it more readable?

Well actually if you get rid of the tuple as I wrote in a different comment this would be solved :)

Reply:
Done.